### DISCUSSION TITLE
VGA interface: How do I go from register documentation to xreg() function calls, and vice versa?

### DISCUSSION THREAD

**Discussion by vruumllc** [picocomputer/community#44] [Q&A] [answered]
OK, I've been staring at the new VGA documentation and vgaprog and mandelbrot code examples for hours, and I am feeling really, really stupid!

I simply cannot go from the VGA register documentation to the associated xreg() function call, or vice versa. The variable parameters xreg() function further confuses the issue, as there are no variable names beyond the first three in the header to give me a clue.  The xreg() calls in vgaprog and mandelbot examples are uncommented, so no clues there either:

// what the ????
    xreg(1, 0, 0, 2);
    xreg(1, 0, 1, 3, 2, 0xFF00, 0, 0, 140);
    xreg(1, 0, 1, 0, 1, 140, 180);

Hopefully this will all be explained in an upcoming video. If any kind soul could enlighten me before then, I would really appreciate it. I'm completely stuck.

**Comment by rumbledethumps** [comment]
The documentation has xreg samples with comments.
```
// Select a 320x240 canvas
result = xreg(1, 0, 0, 1);
// Program mode 3 for 16 colors with
// its config registers at XRAM $FF00.
result = xreg(1, 0, 1, 3, 5, 0xFF00);
```
The variadic is also documented.
  ... – 16 bit integers to set starting at address.

It's not clear what you're stuck on. If you can narrow down the question, there might be an answer.

**Comment by michaelsarr** [comment, marked as answer]
ill give it a shot , 

the first 3 parameters identify the register the 4th is the value
xreg(1, 0, 0, 2); //set register 1:0:0   with the value 2 , CANVAS.  this is 320x180
device 1, channel  0, address  0    set it to 2


// this sets several values at once
// the first 3 parameters identify the device , channel and address 
the 4th ,5th, 6th ... are the values to set... address is incremented
xreg(1, 0, 1, 3, 2, 0xFF00, 0, 0, 140);

device 1, channel  0, address  1 set to 3 (MODE  3 Bitmap)
(since we are mode 3 look up the following from the mode 3 section of the vga documentation)
device 1, channel  0, address  2  set to 2 (OPTIONS  2  - 8 bit color )
device 1, channel  0, address  3 set to  0xFF00 (  CONFIG address of the config structure in ram, other bitmap mode related settings)
device 1, channel  0, address  4  with 0    PLANE , they seem to be 0 is the bottom layer, 1 middle, 2 top ??
device 1, channel  0, address  5   with 0.   scanline begin
device 1, channel  0, address  6   with 140  scanline end

so now plane 0  is mode 3, 320x180 bitmap with 8bit color 



xreg(1, 0, 1, 0, 1, 140, 180);
device 1, channel  0, address  1    set it to 0    MODE  0 is CONSOLE.  so we are enabling the text mode
(look up the following registers under the console section of the vga documentation)
device 1, channel  0, address  2   with 1  PLANE use plane 1, the bitmap defined above is on Plane 0
device 1, channel  0, address  3    with 140 scanline begin
device 1, channel  0, address  4   with 180 scanline end

so now plane 1 is mode 0 console

**Reply by vruumllc** [reply]
THANK YOU! This all makes perfect sense.
The only example in docs (other than CANVAS, which I get), was:

// Program mode 3 for 16 colors with
// its config registers at XRAM $FF00.
result = xreg(1, 0, 1, 3, 5, 0xFF00);

I got completely confused by the 5, for the Bitmap OPTIONS register, documented as:

bit 2 - reverse bits
bit 1:0 - 0=1, 1=4, 2=8, or 3=16 bit color

5 in binary is 101, so that didn't seem to match comment declaring 16 colors. ??? 
It seems like it should be 4 colors with reverse bits (whatever that means).

Also, I was very confused by all the missing arguments (default to ???)

**Reply by vruumllc** [reply]
HEAD SLAP!!! 
I just realized I missed the word 'bit' in the OPTIONS documentation. 4-BIT color is indeed 16 colors, so the comment is correct, although confusing because it is inconsistent (talking about number of color in example vs. number of color bits in table).

**Reply by michaelsarr** [reply]
i have some code to push 4 and 8 bit color images , but I havent gotten 16 bit color to work yet.
the 8 bit color is a bit easier to work with

**Reply by vruumllc** [reply]
Looking forward to checking it out.  Thanks again!

**Comment by netzerohero** [comment]
Going through older posts... the pico-vga area was (and is) challenging for me in working to add simple graphics and text control mode CALL commands to [EhBasic+](https://github.com/picocomputer/ehbasic-plus).  In addition to the [RP6502-VGA docs ](https://picocomputer.github.io/vga.html)and studying tonyvr's [graphics library code](https://github.com/orgs/picocomputer/discussions/53#discussion-5817107) (following michaelsarr's helpful explanation above), I put together a spreadsheet on the pico-vga graphics programming model to aid my understanding. It illustrates:

-  (a) the four canvas-dimensions controlled by 1st vga-pix-register: $1:0:00;  
-  (b) the five canvas-types controlled by the 2nd vga-pix-register: $1:0:01;  
-  Registers 02 through 07 are also illustrated for further configuring the three graphics-planes within the canvas-types.

I'm sharing it here in the hopes it may be of value to others....

[Pico6502_Video_Graphics_Modes_20240211.pdf](https://github.com/picocomputer/community/files/14233557/Pico6502_Video_Graphics_Modes_20240211.pdf)

The capabilities and potential of this humble machine Rumble put together continue to grab my attention....
